pydantic_ai_slim/pydantic_ai/models/openai.py

```diff
@@ -1467,15 +1493,44 @@ def _get_builtin_tools(self, model_request_parameters: ModelRequestParameters) -
             tools.append({'type': 'image_generation'})
         return tools
 
-    def _map_tool_definition(self, f: ToolDefinition) -> responses.FunctionToolParam:
+    def _map_tool_definition(self, f: ToolDefinition) -> responses.FunctionToolParam | responses.CustomToolParam:
+        model_profile = OpenAIModelProfile.from_profile(self.profile)
+        if f.text_format:
+            if not model_profile.openai_supports_freeform_function_calling:
+                raise UserError(
+                    f'Tool {f.name!r} uses freeform function calling but {self._model_name!r} does not support freeform function calling.'
```

Comment:
I'd rather not raise an error, as we won't in other model classes either. We can just use the normal behavior, if possible with agent-side grammar validation